File: aula5/parte1/python/guerreiro.py
from personagem import Personagem


class Guerreiro(Personagem):
    def __init__(self, nome: str, vida: int, nivel: int):
        super().__init__(nome, vida, nivel)
        self.defesa = 5

    @property
    def defesa(self) -> int:
        return self._defesa

    @defesa.setter
    def defesa(self, valor: int) -> None:
        if valor < 0:
            raise ValueError("A defesa não pode ser negativa.")
        self._defesa = valor

    def receber_dano(self, dano: int) -> None:
        dano_efetivo = max(0, dano - self._defesa)
        print(f"{self.nome} bloqueia parte do dano com a defesa ({self._defesa}). "
              f"Dano efetivo: {dano_efetivo}")
        super().receber_dano(dano_efetivo)

    def habilidade(self) -> str:
        return "escudo de aço"

    def ficha(self) -> None:
        super().ficha()
        print(f"Defesa: {self._defesa}")

File: aula5/parte1/java/Guerreiro.java
public class Guerreiro extends Personagem {

    private int defesa;

    public Guerreiro(String nome, int vida, int nivel) {
        super(nome, vida, nivel);
        setDefesa(5);
    }

    public int getDefesa() {
        return this.defesa;
    }

    public void setDefesa(int defesa) {
        if (defesa < 0) {
            throw new IllegalArgumentException("A defesa não pode ser negativa.");
        }
        this.defesa = defesa;
    }

    @Override
    public void receberDano(int dano) {
        int danoEfetivo = Math.max(0, dano - this.defesa);
        System.out.println(this.getNome() + " bloqueia parte do dano com a defesa (" + this.defesa
                + "). Dano efetivo: " + danoEfetivo);
        super.receberDano(danoEfetivo);
    }

    @Override
    public String habilidade() {
        return "escudo de aço";
    }

    @Override
    public void ficha() {
        super.ficha();
        System.out.println("Defesa: " + this.defesa);
    }
}
